types/hooks.d.ts

```diff
@@ -0,0 +1,183 @@
+
+/**
+ * Handler function for onClose hook
+ */
+export type onCloseHookHandler<
+  RawServer extends RawServerBase = RawServerDefault,
+  RawRequest extends RawRequestBase = RawRequestDefault<RawServer>,
+  RawReply extends RawReplyBase = RawReplyDefault<RawServer>
+> = (instance: FastifyInstance<RawServer, RawRequest, RawReply>, done: () => void) => void
+
+/**
+ * Adds a hook that is triggered when a new route is registered. Listeners are passed a
+ * routeOptions object as the sole parameter.
+ * The interface is synchronous, and, as such, the listeners do not get passed a callback.
+ */
+export type onRouteHook<
+  RawServer extends RawServerBase = RawServerDefault,
+  RawRequest extends RawRequestBase = RawRequestDefault<RawServer>,
+  RawReply extends RawReplyBase = RawReplyDefault<RawServer>
+> = (name: 'onRoute', hook: onRouteHookHandler) => FastifyInstance<RawServer, RawRequest, RawReply>
+
+/**
+ * Handler function for onRoute hook
+ */
+export type onRouteHookHandler<
+  RawServer extends RawServerBase = RawServerDefault,
+  RawRequest extends RawRequestBase = RawRequestDefault<RawServer>,
+  RawReply extends RawReplyBase = RawReplyDefault<RawServer>
+> = (opts: RouteOptions<RawServer, RawRequest, RawReply> & { path: string, prefix: string }) => void
+
+/**
+ * Add a hook that is triggered when a request is initially received
+ */
+export type onRequestHook<
+  RawServer extends RawServerBase = RawServerDefault,
+  RawRequest extends RawRequestBase = RawRequestDefault<RawServer>,
+  RawReply extends RawReplyBase = RawReplyDefault<RawServer>
+> = (name: 'onRequest', hook: onRequestHookHandler<RawServer, RawRequest, RawReply>) => FastifyInstance<RawServer, RawRequest, RawReply>
+
+/**
+ * Handler function for onRequest hook
+ */
+export type onRequestHookHandler<
+  RawServer extends RawServerBase = RawServerDefault,
+  RawRequest extends RawRequestBase = RawRequestDefault<RawServer>,
+  RawReply extends RawReplyBase = RawReplyDefault<RawServer>
+> = FastifyMiddleware<RawServer, RawRequest, RawReply>
+
+/**
+ * Hook that is fired after a request is processed, but before the "onResponse"
+ * hook
+ */
+export type onSendHook<
+  RawServer extends RawServerBase = RawServerDefault,
+  RawRequest extends RawRequestBase = RawRequestDefault<RawServer>,
+  RawReply extends RawReplyBase = RawReplyDefault<RawServer>
+> = (name: 'onSend', hook: onSendHookHandler<RawServer, RawRequest, RawReply>) => FastifyInstance<RawServer, RawRequest, RawReply>
+
+/**
+ * Handler function for onSend hook
+ */
+export type onSendHookHandler<
+  RawServer extends RawServerBase = RawServerDefault,
+  RawRequest extends RawRequestBase = RawRequestDefault<RawServer>,
+  RawReply extends RawReplyBase = RawReplyDefault<RawServer>
+> = (
+  req: FastifyRequest<RawServer, RawRequest>,
+  reply: FastifyReply<RawServer, RawReply>,
+  payload: any,
```

Comment:
I have a question. Should we you `any` or `unknown` here ?

Reply:
I'm changing it to unknown. I want to start using unknown more as it make TS users actually implement their typings rather than blindly using any everywhere haha